File: SWEA/D4/1224. ［S／W 문제해결 기본］ 6일차 － 계산기3/［S／W 문제해결 기본］ 6일차 － 계산기3.cpp
#include <bits/stdc++.h>
using namespace std;
stack<char> stc;
stack<int> sti;
map<char,int> m;
 
int main(int argc, char** argv)
{
    ios::sync_with_stdio(0);
    cin.tie(0);
    int test_case;
    m['+']=1;
    m['*']=2;
    for(test_case = 1; test_case <= 10; ++test_case)
    {
        int len;
        cin >> len;
        string infix;
        cin >> infix;
        string postfix;
        // infix -> postfix
        for(int i=0; i<infix.size(); i++) {
            if('0'<=infix[i] && infix[i]<='9') postfix+=infix[i];
            else if(infix[i]=='(') stc.push(infix[i]);
            else if(infix[i]==')') {
            	while(stc.top()!='(') {
                	postfix+=stc.top();
                    stc.pop();
                }
                stc.pop();
            }
            else {
                while(!stc.empty() && m.find(stc.top())->second >= m.find(infix[i])->second) {
                    postfix+=stc.top();
                    stc.pop();
                }
                stc.push(infix[i]);
            }
        }
        while(!stc.empty()) {
            postfix+=stc.top();
            stc.pop();
        }
        // eval postfix
        for(int i=0; i<postfix.size(); i++) {
            if('0'<=postfix[i]&&postfix[i]<='9') sti.push(postfix[i]-'0');
            else {
                int n1 = sti.top();
                sti.pop();
                int n2 = sti.top();
                sti.pop();
                if(postfix[i]=='+') sti.push(n1+n2);
                else sti.push(n1*n2);
            }
        }
        cout << "#" << test_case << " " << sti.top() << '\n';
        sti.pop();
    }
    return 0;
}

File: SWEA/D4/1224. ［S／W 문제해결 기본］ 6일차 － 계산기3/［S／W 문제해결 기본］ 6일차 － 계산기3.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Stack;
import java.util.Map;
import java.util.HashMap;
  
class Solution
{
   public static void main(String args[]) throws Exception
    {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        Stack<Character> stc = new Stack<>();
        Stack<Integer> sti = new Stack<>();
        Map<Character,Integer> map = new HashMap<>();
        map.put('+',1);
        map.put('*',2);
        for(int test_case = 1; test_case <= 10; test_case++)
        {
            int len = Integer.parseInt(br.readLine());
            String input = br.readLine();
            sb.setLength(0);
            // infix -> postfix
            for(int i=0; i<len; i++) {
                if('0'<= input.charAt(i) && input.charAt(i)<='9') sb.append(input.charAt(i));
                else if(input.charAt(i)=='(') stc.push(input.charAt(i));
                else if(input.charAt(i)==')') {
                    char popItem = stc.pop();
                    while(popItem!='(') {
                        sb.append(popItem);
                        popItem=stc.pop();
                    }
                }
                else{
                    while(!stc.isEmpty() && stc.peek()!='(' && map.get(stc.peek())>=map.get(input.charAt(i)) ) sb.append(stc.pop());
                    stc.push(input.charAt(i));
                }
            }
            while(!stc.empty()) sb.append(stc.pop());
            // eval postfix
            for(int i=0; i<sb.length(); i++) {
                if('0'<=sb.charAt(i) && sb.charAt(i)<='9') sti.push(sb.charAt(i)-'0');
                else {
                    int num1 = sti.pop();
                    int num2 = sti.pop();
                    if(sb.charAt(i)=='+') sti.push(num1+num2);
                    else sti.push(num1*num2);
                }
            }
            System.out.println("#"+test_case+" "+sti.pop());
        }
    }
}